lib/src/rules/unreachable_from_main.dart

```diff
@@ -0,0 +1,188 @@
+
+const _desc = 'Unreachable top-level members in executable libraries.';
+
+const _details = r'''
+
+Top-level members in an executable library should be used directly inside this
+library.  Executable libraries are usually never imported and it's better to
+avoid defining unused members.
+
+This rule assumes that an executable library isn't imported by other files
+except to execute its `main` function.
+
+**BAD:**
+
+```dart
+main() {}
+void f() {}
+```
+
+**GOOD:**
+
+```dart
+main() {
+  f();
+}
+void f() {}
+```
+
+''';
+
+class UnreachableFromMain extends LintRule {
+  UnreachableFromMain()
+      : super(
+          name: 'unreachable_from_main',
+          description: _desc,
+          details: _details,
+          group: Group.style,
+          maturity: Maturity.experimental,
+        );
+
+  @override
+  void registerNodeProcessors(
+    NodeLintRegistry registry,
+    LinterContext context,
+  ) {
+    var visitor = _Visitor(this);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule);
+
+  final LintRule rule;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    // TODO(a14n): add support of libs with parts
+    if (node.directives.whereType<PartOfDirective>().isNotEmpty) return;
+    if (node.directives.whereType<PartDirective>().isNotEmpty) return;
+
+    var topDeclarations = node.declarations
+        .expand((e) => [
+              if (e is TopLevelVariableDeclaration)
+                ...e.variables.variables
+              else
+                e,
+            ])
+        .toSet();
+
+    var entryPoints = topDeclarations.where(_isEntryPoint).toList();
+    if (entryPoints.isEmpty) return;
+
+    var declarationByElement = <Element, Declaration>{};
+    for (var declaration in topDeclarations) {
+      var element = declaration.declaredElement;
+      if (element != null) {
+        if (element is TopLevelVariableElement) {
+          declarationByElement[element] = declaration;
+          var getter = element.getter;
+          if (getter != null) declarationByElement[getter] = declaration;
+          var setter = element.setter;
+          if (setter != null) declarationByElement[setter] = declaration;
+        } else {
+          declarationByElement[element] = declaration;
+        }
+      }
+    }
+
+    var dependencies = Map<Declaration, Set<Declaration>>.fromIterable(
+      topDeclarations,
+      value: (declaration) =>
+          DartTypeUtilities.traverseNodesInDFS(declaration as Declaration)
+              .expand((e) => [
+                    if (e is SimpleIdentifier) e.staticElement,
+                    // with `id++` staticElement of `id` is null
+                    if (e is CompoundAssignmentExpression) ...[
+                      e.readElement,
+                      e.writeElement,
+                    ],
+                  ])
+              .whereNotNull()
+              .map((e) {
```

Comment:
I think this `map` callback is equivalent to some invocation of `Element.thisOrAncestorMatching`, which would make it more readable. Something like

```dart
.map((e) =>
    e.thisOrAncestorMatching(
        (a) => a.enclosingElement2 == null || a.enclosingElement2 is CompilationUnitElement))
```

maybe? WDYT?

Reply:
done